torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -198,28 +240,31 @@ void PropagateShapeOnNode(Node * node, bool insert_expands) {
       }
       return;
     }
-    case prim::NumToTensor:
-    case prim::TensorToNum: {
-      node->output()->setType(node->inputs()[0]->type());
-      return;
-    }
+    case prim::TensorToNum:
+      return; // correct num type is already set
     case prim::Constant: {
-      node->output()->inferTypeFrom(node->t(attr::value));
+      if(node->output()->type()->isSubtypeOf(*DynamicType::get())) {
+        node->output()->inferTypeFrom(node->t(attr::value));
+      }
```

Comment:
Otherwise, we're guaranteed to have the correct type set, right?

Reply:
yes, the only other constants are int/float